### DISCUSSION TITLE
布局的问题

### DISCUSSION THREAD

**Discussion by rose8928** [opensumi/core#2107] [Q&A]
比如新建一个lua文件，我想在正常编辑器里面第一行代码的上面和面包屑breadcrumb-menu的下面，添加一行我自己的组件（类似于工具栏），要使用什么服务或者方式。

**Comment by erha19** [comment]
@rose8928 顶部区域暂时不支持通过代码拓展添加自定义组件，目前仅支持在编辑器底部追加自定义组件，代码如：

https://github.com/opensumi/core/blob/9af5e4fbc53bbb3d76ce88b88b07817d09b941fd/packages/startup/src/browser/startup.contribution.ts#L65

效果如下：
<img width="889" alt="image" src="https://user-images.githubusercontent.com/9823838/208880918-f54c8728-8256-47b3-8413-82d10d2e374b.png">

如果不满足需求，可以考虑来我们仓库贡献代码支持一下顶部区域拓展。

**Comment by rose8928** [comment]
好的，感谢，如果我想根据我打开的文件，添加一个搜索或者命令这种悬浮组件，使用什么服务或者方法，类似于下图这种：
![float1](https://user-images.githubusercontent.com/35814595/209039369-00555f05-4007-4185-8c85-6ed45dbe9339.png)
![float2](https://user-images.githubusercontent.com/35814595/209039384-f768f91d-3fb1-463a-87f1-7d261a488d00.png)

**Reply by erha19** [reply]
可以用一下 `IQuickInputService`.

**Reply by rose8928** [reply]
> 可以用一下 `IQuickInputService`.

我是想创建一个悬浮组件，组件我自己可以随意设计，悬浮于editor上面，里面可以没有input，或者说是类似于面包屑
[SlotLocation.extra]: {
modules: ['breadcrumb-menu'],
},
这种的功能。如何实现。
另外还想问一下SlotLocation.float是如何使用的

**Reply by erha19** [reply]
自定义布局可以看一下文档
https://opensumi.com/zh/docs/integrate/universal-integrate-case/custom-view

也有参考的案例，如拓展 Toolbar 的案例
https://github.com/opensumi/opensumi-module-samples/tree/main/modules/custom-toolbar

**Reply by rose8928** [reply]
非常感谢，下午学习了一下很有用。还有一个问题就是，我可以动态的隐藏和显示我的某个布局slotlocation吗，有这种服务吗？比如，可以隐藏bottom和left的子视图，但是没有找到可以隐藏整个bottom或者left的服务，除非开始就不注册上去。

**Reply by erha19** [reply]
@rose8928  看一下这个例子 https://github.com/opensumi/opensumi-modue-samples/blob/main/modules/toggle-panel/browser/toggle-panel.contribution.ts

**Comment by rose8928** [comment]
this._mainLayoutService.getTabbarService(SlotLocation.bottom).updatePanelVisibility(false);试了这种方法，可以隐藏slotlocation.left, slotlocation.bottom，但是无法隐藏我自定义的插槽customaction，无法隐藏slotlocation.top，是吗？自定义layout是这样加的
![customlayout](https://user-images.githubusercontent.com/35814595/209279466-872b86ca-3d86-43aa-a3db-1216883322a4.png)

**Reply by erha19** [reply]
@rose8928 目前具备 “开关” 属性的面板只有“左右下”面板，对于自定义的面板隐藏，可以通过进入时修改布局模板的方式隐藏，也可以自己控制样式去隐藏，目前不具备 API 实现该能力。
